internal/ingress/controller/controller.go

```diff
@@ -591,11 +591,13 @@ func (n *NGINXController) getBackendServers(ingresses []*ingress.Ingress) ([]*in
 									location.Path, server.Hostname, location.DefaultBackend.Namespace, location.DefaultBackend.Name)
 
 								nb := upstream.DeepCopy()
-								name := fmt.Sprintf("custom-default-backend-%v", upstream.Name)
+								name := fmt.Sprintf("custom-default-backend-%v", location.DefaultBackend.GetName())
```

Comment:
This change means that the backend name of custom-error locations is `custom-default-backend-<service_name>` where `service_name` is the name of the default-backend service. It used to be that the `service_name` would be the name of the upstream for which this default-backend is substituted. Eg. where it used to be `custom-default-backend-my-app`, it is now `custom-default-backend-app-errorhandler`.

This is necessary since multiple locations within a server block can share the same 'default-backend' - so now, the name of the error-handler-location clearly signifies which service it will route to.

Reply:
Is this a necessary change? I wonder what was the rationale to design that way initially? Could it be maybe because it would explicitly mention what upstream it is replacing due to the fact that upstream has no endpoints?

Reply:
well it's necessary if multiple ingress objects (ie locations) for a single upstream can have different error blocks. if I don't make this change, all locations that point to a given upstream have to share the same default-backend.

I assume it was that way initially since the default-backend was only for when the upstream had no endpoints, thus this upstream was replacing the regular upstream entirely (whereas now, if endpoints > 0, the default backend is just another upstream).

Reply:
Given two ingresses one with location /one pointing to upstream X and another with location /two pointing to upstream X as well. And upstream X has no endpoints. Then without this change we can not use different `default-backend` annotation per location, right?

Given that this change is not necessarily about custom errors, do you think we should introduce it in a different PR with its own tests and documentation adjustments?

Reply:
the case I was more targeting is two ingresses /one and /two, same upstream, both with custom error annotations but different default backends - but yeah, that case you mentioned is also true.

Reply:
It might be tricky to pull it out into another PR - I'll look and see how tightly I integrated the rest. I don't think it will have any user-facing change, but we can definitely add tests and update the docs (https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#default-backend is out-of-date anyway)